tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java

```diff
@@ -0,0 +1,245 @@
+ * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * Contributors:
+ *   Andrew Helwer - code interfacing with SANY
+ *   Markus Alexander Kuppe - code interfacing with TLC interpreter
+ ******************************************************************************/
+package tlc2.debug;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import tla2sany.parser.SyntaxTreeNode;
+import tla2sany.parser.TLAplusParser;
+import tla2sany.semantic.AbortException;
+import tla2sany.semantic.Errors;
+import tla2sany.semantic.FormalParamNode;
+import tla2sany.semantic.Generator;
+import tla2sany.semantic.LetInNode;
+import tla2sany.semantic.LevelConstants;
+import tla2sany.semantic.ModuleNode;
+import tla2sany.semantic.OpApplNode;
+import tla2sany.semantic.OpArgNode;
+import tla2sany.semantic.OpDefNode;
+import tla2sany.semantic.SemanticNode;
+import tla2sany.semantic.SymbolNode;
+import tla2sany.st.Location;
+import tlc2.tool.impl.SpecProcessor;
+import util.ToolIO;
+
+/**
+ * An expression for conditionally triggering a breakpoint.
+ */
+public class TLCBreakpointExpression {
+
+	/**
+	 * Given a spec and an unparsed expression, build an operator that can be
+	 * evaluated within the context of that spec.
+	 *
+	 * @param processor     Processes expression within the model context.
+	 * @param semanticRoot  The root of the spec's semantic parse tree.
+	 * @param location      The breakpoint location.
+	 * @param conditionExpr The unparsed breakpoint expression.
+	 * @return A breakpoint expression, or null if parsing failed.
+	 */
+	public static OpDefNode process(
+			final SpecProcessor processor,
+			final ModuleNode semanticRoot,
+			final Location location,
+			final String conditionExpr
+	) {
+		if (null == conditionExpr || conditionExpr.isBlank()) {
+			return null;
+		}
+
+		final String rootModName = semanticRoot.getName().toString();
+		final String bpModName = semanticRoot.generateUnusedName("__BreakpointModule__%s");
+		final String bpOpName = semanticRoot.generateUnusedName("__BreakpointExpr__%s");
+		final Set<String> paramNames = getScopedIdentifiers(semanticRoot, location);
+		final String params = paramNames.size() > 0 ? "(" + String.join(", ", paramNames) + ")" : "";
+		final String bpOpDef = bpOpName + params;
+
+		final String wrapper = "---- MODULE %s ----\nEXTENDS %s\n%s == %s\n====";
+		String wrappedConditionExpr = String.format(wrapper, bpModName, rootModName, bpOpDef, conditionExpr);
+		byte[] wrappedConditionExprBytes = wrappedConditionExpr.getBytes(StandardCharsets.UTF_8);
+		InputStream sourceCode = new ByteArrayInputStream(wrappedConditionExprBytes);
+		TLAplusParser parser = new TLAplusParser(sourceCode, StandardCharsets.UTF_8.name());
+		boolean syntaxParseSuccess = parser.parse();
```

Comment:
A `conditionExpr` such as "Does not exist" cause `parse()` to print:

```
***Parse Error***
Was expecting "==== or more Module body"
Encountered "not" at line 3, column 39 and token "Does" 
```

Reply:
See `EWD998ChanDebuggerTest` after cherry-picking https://github.com/tlaplus/tlaplus/commit/8baf0d86ac3d14f26c11c3b3261a6b8f2e70eb54

Reply:
To alleviate the issue, the error message could be modified to include the module name, thereby prompting users to recognize that the problem lies within the debug module.

Reply:
Added some additional error logging